### PR TITLE
image picker: select an icon previously uploaded by the organization admin

### PR DESCRIPTION
**Pull request by matallo**
fix https://github.com/CartoDB/cartodb/issues/11462

added the organization uploads tab to the image picker

a7f44640-9c81-44b9-8f12-5e2810304d55

I didn't cover the case of icon upload by the organization admin, as I think it fails out of the scope of the issue, but it could be extended afterwards to support this (and better resemble the "Your icons" tabs)

---

**acceptance**

- add an icon as an organization admin
- and use it in the map
- if the user doesn't belong to an organization, it shouldn't appear at all

### CODE REVIEW

**Comment by matallo** [comment]
quite advanced on this, do we have an state for the case when the organization doesn't have any icon uploaded yet? @xavijam @javierarce 

b5e6ff68-4a4f-4d6e-aa99-ee8c53eb9858

the first thing that came to my mind was not showing the tab at all, but it is after the modal is rendered that the organization icons are fetched :/

**Comment by xavijam** [comment]
That's a good one, don't we have a 0 results state? have you seen it in the designs @javierarce?

**Comment by javierarce** [comment]
No, I'm afraid we don't have it / I haven't seen it.

**Comment by xavijam** [comment]
Let's go to the source of truth, design 🍡 .

**Comment by javierarce** [comment]
I guess hiding the tab in case the user didn't have any assets would be the best solution, but that would require an extra call.

**Comment by matallo** [comment]
if we can afford that extra call I'm down to follow that path of thinking

**Comment by xavijam** [comment]
Fine by me, go ahead!

**Comment by matallo** [comment]
now the tab would only appear if the User is an Organization user and the organization icons collection is not empty

we need to make two extra calls (the organization icons collection, and the user icons collection, which we had to extract) but further requests are cached

File: lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js [hunk 1]
```diff
@@ -352,8 +404,22 @@ module.exports = CoreView.extend({
...
+  _onFetchOrganizationAssetsError: function (model, response) {
+    this._setError(response);
+  },
+
+  _onFetchUserAssetsError: function (model, response) {
+    this._setError(response);
+  },
+
+  _onFetchOrganizationAssets: function () {
+    this._stateModel.unset('status');
```
Comment (nobuti) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: I feel this weird, I understand it but it's not natural to me. Also, this led to having two render calls, if I'm not wrong.

Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js]: Indeed, let's open for discussion, there's a listen with the model change which renders the view.

https://github.com/CartoDB/cartodb/blob/b7409461881839b582d170ca608fe568317dee58/lib/assets/core/javascripts/cartodb3/components/form-components/editors/fill/input-color/assets-picker/assets-view.js#L114

I just tried to follow the current flow. Any suggestion to improve this?

**Comment by xavijam** [comment]
I encourage you how, for example, we code the add widget modal, where several requests are needed before render the view displaying a new state. For me it should work in the same way, if possible (take into account this modal will be the main modal in the future, so better to make it properly than other thing).

**Comment by matallo** [comment]
roger!

**Comment by matallo** [comment]
I tried to go full widgets implementation, but there's no model related to fetched collections to rely on, and after replicating the `bodyView` and that stuff I trimmed back to something quite similar

anyways, I took into account the `$.when` suggestion (and fixed the tests)

https://github.com/CartoDB/cartodb/pull/11492/files#diff-f6ad32d8aa3da552f4ab34d6b38853a1R114

**Comment by xavijam** [comment]
Hey, I suggested you to not follow the same solution, just take it as an example :).
Question, if the user doesn't belong to an organization, organization icons fetch fails or it returns an empty array? I guess we don't show the error state in that case right?

**Comment by matallo** [comment]
no prob! my bad for trying too far. If the user doesn't belong to the organization, the array of collections to `apply` fetch will only contain`userAssetCollection`, thus no errors and only the "your uploads" tab will be shown

(solving the conflicts now to move on)

**Comment by xavijam** [comment]
:+1:

**Comment by javierarce** [comment]
I went ahed and included the change for the `findWhere`. Feel free to act on the other two comments :)

**Comment by xavijam** [comment]
Nice one, let's go to acceptance?

**Comment by javierarce** [comment]
### Acceptance

I've detected a (probably small) bug:

1. An organization uploads a bunch of icons.
2. A member of the organization selects an icon to style a marker.
3. The organization removes that particular icon from the ones it offers.

If that happens and the user goes back to the icon modal and to the organization tab, the "set button" will be enabled, but there won't be any icon selected.

There's also the issue of having a map with an icon that no longer exists (only in the cache).

**Comment by xavijam** [comment]
About the icon removed and how the fill component will look like, it is a good question. @CartoDB/design what do you think?
Further to the last "question", it is something already discussed and there is a big and great warning in the organization side. We decided to not do it nothing else for the moment (https://github.com/CartoDB/cartodb/pull/11292#issuecomment-273751733).

**Comment by saleiva** [comment]
@noguerol's call

**Comment by noguerol** [comment]
Can't we just go back to a default bubble whenever a map with missing icon is loaded?

**Comment by javierarce** [comment]
@noguerol I don't think that's an easy thing to do right now, since we'd need to store the previous size somewhere.

**Comment by javierarce** [comment]
### Acceptance 

I've discovered a big issue with this PR. It seems that uploading a custom icon as a regular user, doesn't refresh the list of icons in the modal :(

**Comment by matallo** [comment]
d'oh, very well spotted as I think I only tested this as owner :((( move to development then andI'll take as soon as I can

**Comment by javierarce** [comment]
Oh, don't worry, I'll fix it and ask you for help if I need it.

**Comment by javierarce** [comment]
Back to CR! I've addressed the problems ([1](https://github.com/CartoDB/cartodb/pull/11492#issuecomment-279968180) and [2](https://github.com/CartoDB/cartodb/pull/11492#issuecomment-280019341)) found in Acceptance.

Props to @matallo for the help me debugging one of the problems.

**Comment by xavijam** [comment]
Taking a look @javierarce 💋

**Comment by noguerol** [comment]
OK, so if found one little issue and one thing that concerns me:

a, the issue: on the organizational settings, the copy for deleting an icon is in plural ("delete icons"), while it only deletes one icon, so is confusing as it seems it will erase them all

b. the concern: I understand the dont-break-maps policy when removing an organizational because is very difficult to know how many maps use them, but I can't find a reason to still render a map with an icon that has been deleted when its opened on Builder. On that situation I think we can afford to do it and will be better whatever the use case is.

**Comment by xavijam** [comment]
@noguerol 

a. It is not related with this PR (check description of this pull request), if you consider it is confusing, please open a new ticket.
b. we already discuss that and we took a decission. Weren't you in that thread?. What's your proposal?

**Comment by noguerol** [comment]
Understood. Opening two new tickets, lets move this to production.

**Comment by xavijam** [comment]
💋

**Comment by matallo** [comment]
merging and bumping assets in master

**Pull request merged**